Fix Windows Installation Bugs & Non-numeric installs

rhass-r7 opened the pull request:
# Summary

This fixes a number of bugs with Windows installs, and one major bug which affected all platforms.
# Windows
- Fixed bug in which only one version off Firefox could be installed at a time. When multiple versions were specified, only the last to install would actually be installed.
  - Fixed bug in which named versions (latest, latest-esr) were not idempotent on Windows as the version string used to check the registry key for the windows_package resource was incorrect.
# All Platforms
- Fixed bug in which named releases would never install newer versions than the originally installed version due to use of create_if_missing action in remote_file resource.
- Added Chef Auditing recipe for integration testing. This is mostly an Ubuntu test suite as I had issues with ServerSpec for Windows with be_executable and be_version.
# Linux
- Added Travis CI support.
# Validation
- [ ] Verify Travis build.
- [ ] Run 'kitchen verify' to validate Windows, and check to see if all three Firefox versions installed. Note: The vagrant box in the kitchen.yml for Windows had usb enabled, which caused it nut to bit in my virtualbox, dissing this resolved the issue.
# Suggested Reviewers

@bturner-r7 @cdoughty-r7 @asebastian-r7

rhass-r7 committed:
Add Travis Support

This is an attempt to use Docker via travis. With their recent switch
to docker this may work now, or it may not. This is an initial attempt
to find out.

rhass-r7 committed:
Download the docker binary

rhass-r7 committed:
Rock travis-ci with some docker magic.

I found a stackoverflow describing how to dockerfy travis builds.
http://stackoverflow.com/questions/26098531/can-travis-ci-run-docker

rhass-r7 committed:
Create test cookbook for integration testing.

The test cookbook current comprises of two recipes, the default recipe
and the audit mode controls recipe. The latter of the two leverages
Chef Audit Mode to run ServerSpec inline without having to use busser.
This may, or may not be the right approach to testing cookbooks later
on, but it seems to be the direction in which Chef is taking things.

rhass-r7 committed:
Do not use travis container based infrastructure.

While it is technically possible to run docker within docker, I have
not had any previous success in doing so. This will ensure we can run
docker based builds for testing this cookbook.

rhass-r7 committed:
Make apt-get update chatty to see error in travis.

rhass-r7 committed:
Try the sudo curl thing.

Because curling random shell scripts into bash is always safe.

rhass-r7 committed:
Fix SLIRP deprecation warning.

The last build hung and before doing so threw a SLIRP deprecation
warning. It's not entirely likely the two are related, however I found
some other projects with this newer setup which are building correctly.

rhass-r7 committed:
Add Windows support to Test Kitchen

rhass-r7 committed:
Committing for review and debug help.

rhass-r7 committed:
Merge branch 'master' into windows_install_fix-debug

Conflicts:
	.kitchen.travis.yml
	.kitchen.yml
	.travis.yml
	test/cookbooks/firefox_package_test/metadata.rb
	test/cookbooks/firefox_package_test/recipes/default.rb
	test/cookbooks/firefox_package_test/recipes/default_controls.rb

rhass-r7 committed:
Fixed Windows and latest release installs.

The Windows install would only allow for a single version to be
installed at a time. This now allows multiple versions to be installed.
Moreover, the latest and latest-esr values were broken for all
platforms as it would never upgrade the version past the initially
installed version, it will now upgrade the version automatically and
handle installation detection on Windows correctly.

erran-r7 commented on libraries/firefox_package.rb:
```diff
@@ -5,7 +5,7 @@
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
+# you may not us e this file except in compliance with the License.
```
Typo

erran-r7 commented on libraries/firefox_package.rb:
```diff
@@ -74,33 +77,72 @@ def munged_platform
+
+
+    def esr?(filename)
+      version = /esr/.match(filename)
```
Prefer `filename =~ /esr/` if you're ignoring the return value. That will give a truthy result if the esr was within the filename and falsy if it was not.

erran-r7 replied:
To be clear, I meant ignoring the match data. All you were doing is checking for the presence. If you don't need the match data object we can skip initializing a new one.

erran-r7 commented on test/cookbooks/firefox_package_test/recipes/default_controls.rb:
```diff
@@ -1,43 +1,78 @@
+      end
+
+# This seems to be broken
+#      it 'is the correct version of Firefox'do
```
You should be able to use `xit` instead of `it`. That way you can leave the test committed but it will be marked as pending.

rhass-r7 committed:
Fix boot failure for Windows in VirutalBox.

The base box we are using has a USB option enabled which fails on my
machine. Since we really do not care about USB capability in this
context, we are disabling it as a virtualbox config option. This
ensures the system boots correctly every time.

rhass-r7 committed:
Fix various small issues identified in code review

rhass-r7 committed:
Use seconds instead of minutes for splay.

The standard for time delay is usually seconds, and I found myself
using the incorrect values. As such, I have changed this to seconds.
